lib/util.js

```diff
@@ -241,6 +243,23 @@ function inspectPromise(p) {
 
 
 function formatValue(ctx, value, recurseTimes) {
+
+  if (ctx.showProxy && typeof value === 'object') {
+    const inProxyCache = ctx.proxyCache.has(value);
+    var proxy = inProxyCache ? ctx.proxyCache.get(value) : undefined;
+    if (inProxyCache && proxy) {
+      return 'Proxy ' + formatValue(ctx, proxy, recurseTimes);
+    } else if (!inProxyCache) {
+      if (binding.isProxy(value)) {
+        proxy = binding.getProxyDetails(value);
+        ctx.proxyCache.set(value, proxy);
+        return 'Proxy ' + formatValue(ctx, proxy, recurseTimes);
+      } else {
+        ctx.proxyCache.set(value, undefined);
```

Comment:
Unfortunately, no. This is here to record the object's that we know *aren't* proxies in case we encounter them again while we are recursing through. This is why I do the `has` check above before the `get` also. Consider:

```js
var a = new Proxy({}, {});
var b = { };
var c = [ b, a, b, a ];
```

When I pass this to `util.inspect()`, util.inspect does not have any idea which objects are Proxies and which are normal objects until it does the isProxy call, which is expensive. If formatValue has already checked `b` above and has determined that it's not a Proxy, if we don't record that knowledge somehow we end up having to do the check again when `b` comes back around. This is largely because `util.inspect()` is not very efficient at caching stuff that's already seen. There's some memoization happening later on that I can see about refactoring, but without recording that we've already ruled an object out, this ends up being much slower.